pkg/util/tlsutil/x509.go

```diff
@@ -0,0 +1,149 @@
+package tlsutil
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"time"
+)
+
+var (
+	Duration365d = time.Hour * 24 * 365
+)
+
+const (
+	RSAKeySize = 2048
+)
+
+type CACertConfig struct {
+	CommonName   string
+	Organization string
+	Duration     time.Duration
+	PublicKey    crypto.PublicKey
+}
+
+type CertConfig struct {
+	CommonName   string
+	Organization string
+	Duration     time.Duration
+	DNSNames     []string
+	IPAddresses  []string
+	PublicKey    crypto.PublicKey
+	extKeyUsage  []x509.ExtKeyUsage
+}
+
+func NewPrivateKey() (*rsa.PrivateKey, error) {
+	privKey, err := rsa.GenerateKey(rand.Reader, RSAKeySize)
+	if err != nil {
+		return nil, fmt.Errorf("error generating RSA private key: %v", err)
+	}
+	return privKey, nil
+}
+
+//Certificate Signers (things which mint identities. Pulling the interface spec out now will make implementing new ones cake)
+
+// This is explictly not a certificate request model
+// ie: the singee must divulge private key material to singer
+// This is fine for a one-way trust delegation model (CA downwards)
+type CertificateSigner func(pubKey crypto.PublicKey, certTmpl *x509.Certificate) (*x509.Certificate, error)
+
+// We have the ca private key in memory already
```

Comment:
The `Local` bit already explains that, right? If we want to make it more clear, can we make this comment part of the func comment of `NewLocalCertificateSigner `?